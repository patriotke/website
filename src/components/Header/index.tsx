/**
 *
 * Header
 *
 */
import Link from 'next/link';
import { clsx } from 'clsx';

export type ILink = {
  href: string;
  text: string;
  active: boolean;
  id: string;
};
export type IHeader = {
  links: ILink[];
};

function Header({ links = [] }: IHeader) {
  return (
    <header className="flex flex-wrap sm:justify-start sm:flex-nowrap w-full bg-white text-sm py-4 border-y border-gray-200 sticky top-0">
      <nav
        className="max-w-[85rem] w-full mx-auto px-4 sm:flex sm:items-center sm:justify-between"
        aria-label="Global"
      >
        <Link className="flex-none text-xl font-semibold" href="/">
          {process.env.NEXT_PUBLIC_TITLE}
        </Link>
        <div className="flex flex-row items-center gap-5 mt-5 sm:justify-end sm:mt-0 sm:ps-5">
          {links.map((link) => (
            <Link
              key={link.id}
              href={link.href}
              className={clsx('font-medium', {
                'text-blue-500': link.active,
                'text-gray-600 hover:text-gray-900': !link.active,
              })}
            >
              {link.text}
            </Link>
          ))}
        </div>
      </nav>
    </header>
  );
}

export default Header;
