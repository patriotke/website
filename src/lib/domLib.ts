// eslint-disable-next-line import/prefer-default-export
export const someFn = () => {
  console.log('someFn');
  // do something
  return 'someFn';
};

const getJobs = (table) => {
  const jobs = [];
  // eslint-disable-next-line no-plusplus
  for (let r = 0, n = table.rows.length; r < n; r++) {
    // eslint-disable-next-line no-plusplus
    for (let c = 0, m = table.rows[r].cells.length; c < m; c++) {
      if (c === 3) jobs.push(table.rows[r].cells[c].innerText.toLowerCase());
    }
  }
  return jobs;
};

const groupCommonJobs = (jobs) => {};

const summarizeJobs = (jobs: string[]) =>
  jobs.reduce(
    (acc, job) => {
      if (acc[job]) {
        acc[job] += 1;
      } else {
        acc[job] = 1;
      }
      return acc;
    },
    {} as { [key: string]: number },
  );
