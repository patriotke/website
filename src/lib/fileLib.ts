// eslint-disable-next-line import/prefer-default-export
export const getFileExtensionFomMimeType = (mimeType: string): string => {
  console.log(mimeType);
  // do something
  return 'someFn';
};
