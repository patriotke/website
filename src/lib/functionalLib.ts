import { Maybe } from 'monio';

// eslint-disable-next-line import/prefer-default-export
export const getPropSafe = (prop: string) => (obj: object) =>
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-expect-error
  Maybe.from(obj[prop]);
