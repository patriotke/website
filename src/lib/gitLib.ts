// eslint-disable-next-line import/no-extraneous-dependencies
import { Octokit } from 'octokit';

export const getOctokit = () =>
  new Octokit({
    auth: process.env.GITHUB_TOKEN!,
  });

/**
 * List branches
 */
export const listBranches = async () => {
  const url = `https://api.github.com/repos/${process.env.GITHUB_OWNER}/${process.env.GITHUB_REPO}/branches`;
  return fetch(url, {
    headers: {
      Authorization: `token ${process.env.GITHUB_TOKEN}`,
    },
  }).then((res) => res.json());
};

/**
 * Get branch
 * @param branchName
 */
export const getBranch = async (branchName: string) => {
  const url = `https://api.github.com/repos/${process.env.GITHUB_OWNER}/${process.env.GITHUB_REPO}/branches/${branchName}`;
  return fetch(url, {
    next: { revalidate: 0 },
    headers: {
      Authorization: `token ${process.env.GITHUB_TOKEN}`,
    },
  }).then((res) => res.json());
};

/**
 * Create branch
 * @param branchName
 */
export const createBranch = async (branchName: string) => {
  const octokit = getOctokit();
  const sha = await getBranch('main').then((res) => res.commit.sha);
  // console.log('sha', sha);
  return octokit.request('POST /repos/{owner}/{repo}/git/refs', {
    sha,
    owner: process.env.GITHUB_OWNER!,
    repo: process.env.GITHUB_REPO!,
    ref: `refs/heads/${branchName}`,
  });
};

/**
 * Get blob
 * @param branchName
 * @param path
 */
export const getBlob = async (branchName: string, path: string) => {
  const url = `https://api.github.com/repos/${process.env.GITHUB_OWNER}/${process.env.GITHUB_REPO}/contents/${path}?ref=${branchName}`;
  return fetch(url, {
    next: { revalidate: 0 },
    headers: {
      Authorization: `token ${process.env.GITHUB_TOKEN}`,
    },
  }).then((res) => res.json());
};

/**
 * Commit change
 * @param branchName
 * @param path
 * @param content
 */
export const commitChange = async (
  branchName: string,
  path: string,
  content: string | ArrayBuffer,
) => {
  const octokit = getOctokit();
  const sha = await getBlob(branchName, path)
    .then((res) => res.sha)
    .catch(() => getBranch(branchName).then((res) => res.commit.sha));
  // console.log('sha', sha);
  return octokit.request('PUT /repos/{owner}/{repo}/contents/{path}', {
    owner: process.env.GITHUB_OWNER!,
    repo: process.env.GITHUB_REPO!,
    path,
    message: `Add ${path} ${new Date().toISOString()}`,
    content: Buffer.from(content).toString('base64'),
    branch: branchName,
    sha,
  });
};

/**
 * Create pull request
 * @param branchName
 */
export const createPullRequest = async (branchName: string) => {
  const octokit = getOctokit();
  return octokit.request('POST /repos/{owner}/{repo}/pulls', {
    owner: process.env.GITHUB_OWNER!,
    repo: process.env.GITHUB_REPO!,
    title: `Add ${branchName}`,
    body: `Add ${branchName}`,
    head: branchName,
    base: 'main',
  });
};
