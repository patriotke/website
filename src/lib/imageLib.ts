/**
 * Get image preview from file
 * @param file
 */
// eslint-disable-next-line import/prefer-default-export
export const getImagePreview = (file: File): HTMLImageElement => {
  const image = document.createElement('img');
  image.src = URL.createObjectURL(file);
  image.alt = file.name;
  image.title = file.name;
  image.className = 'aspect-square object-cover';
  return image;
};

/**
 * Create preview
 * @param parent
 * @param image
 */
export const createPreview = (
  parent: HTMLDivElement,
  image: HTMLImageElement,
) => parent.replaceChildren(image);

/**
 * Get file base64
 * @param file
 */
export const getFileBase64 = (file: File): Promise<string> =>
  new Promise((resolve) => {
    const reader = new FileReader();
    reader.onload = () => resolve(reader.result as string);
    reader.readAsDataURL(file);
  });

/**
 * Get blob from data url
 * @param dataUrl
 */
export const getBlobFromDataUrl = (dataUrl: string): Blob => {
  const [mimeContainer, b64] = dataUrl.split(',');
  const [mime] = mimeContainer.split(':')[1].split(';');
  const u8arr = new Uint8Array(
    atob(b64)
      .split('')
      .map((c) => c.charCodeAt(0)),
  );
  return new Blob([u8arr], { type: mime });
};
