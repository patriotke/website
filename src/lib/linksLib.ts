import { ILink } from '@/types/header.types';

// eslint-disable-next-line import/prefer-default-export
export const getHeaderLinks = (): ILink[] => [
  {
    href: '/about',
    text: 'About',
    id: 'about',
    active: false,
  },
  {
    href: '/contribute',
    text: 'Contribute',
    id: 'contribute',
    active: false,
  },
];

export const getStatefulLinks =
  (active: string) =>
  (links: ILink[]): ILink[] =>
    links.map((link) => ({ ...link, active: link.href === active }));
