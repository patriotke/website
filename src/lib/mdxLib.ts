// eslint-disable-next-line import/prefer-default-export
export const someFn = () => {
  console.log('someFn');
  // do something
  return 'someFn';
};
