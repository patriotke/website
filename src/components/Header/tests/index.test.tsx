/* eslint-disable import/no-extraneous-dependencies */
import '@testing-library/jest-dom';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { render, screen } from '@testing-library/react';
import { getHeaderLinks } from '@/lib/linksLib';
import Header from '../index';

describe('Header', () => {
  it('renders Header', () => {
    const { container } = render(<Header links={getHeaderLinks()} />);
    expect(container).toMatchSnapshot();
  });

  it('render Header with links', () => {
    const link = getHeaderLinks()[0];
    render(<Header links={getHeaderLinks()} />);
    expect(screen.getByText(link.text).closest('a')).toHaveAttribute(
      'href',
      link.href,
    );
    expect(screen.getByText('App').closest('a')).toHaveAttribute('href', '/');
  });
});
