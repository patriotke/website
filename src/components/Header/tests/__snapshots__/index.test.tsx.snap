// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders Header 1`] = `
<div>
  <header
    class="flex flex-wrap sm:justify-start sm:flex-nowrap w-full bg-white text-sm py-4 border-y border-gray-200"
  >
    <nav
      aria-label="Global"
      class="max-w-[85rem] w-full mx-auto px-4 sm:flex sm:items-center sm:justify-between"
    >
      <a
        class="flex-none text-xl font-semibold"
        href="/"
      >
        App
      </a>
      <div
        class="flex flex-row items-center gap-5 mt-5 sm:justify-end sm:mt-0 sm:ps-5"
      >
        <a
          class="font-medium text-gray-600 hover:text-gray-900"
          href="/about"
        >
          About
        </a>
        <a
          class="font-medium text-gray-600 hover:text-gray-900"
          href="/contribute"
        >
          Contribute
        </a>
      </div>
    </nav>
  </header>
</div>
`;
